File: src/Pudicitia.Identity.Api/Services/AuthorizationService.cs
using Pudicitia.Identity.App.Authorization;

namespace Pudicitia.Identity.Api;

public class AuthorizationService : Authorization.AuthorizationBase
{
    private readonly AuthorizationApp _authorizationApp;

    public AuthorizationService(AuthorizationApp authorizationApp)
    {
        _authorizationApp = authorizationApp ?? throw new ArgumentNullException(nameof(authorizationApp));
    }

    public override async Task<ListNamedEntityResponse> ListUsers(
        ListUsersRequest request,
        ServerCallContext context)
    {
        var users = await _authorizationApp.GetUsersAsync(request.UserName);
        var items = users.Select(x => new ListNamedEntityResponse.Types.NamedEntity
        {
            Id = x.Id,
            Name = x.Name,
        });
        var result = new ListNamedEntityResponse();
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<PaginateUsersResponse> PaginateUsers(
        PaginateUsersRequest request,
        ServerCallContext context)
    {
        var options = new UserOptions
        {
            Page = request.Page,
            UserName = request.UserName,
            Name = request.Name,
            IsEnabled = request.IsEnabled,
        };

        var users = await _authorizationApp.GetUsersAsync(options);
        var items = users.Items.Select(x => new PaginateUsersResponse.Types.User
        {
            Id = x.Id,
            UserName = x.UserName,
            Name = x.Name,
            DisplayName = x.DisplayName,
            IsEnabled = x.IsEnabled,
        });
        var result = new PaginateUsersResponse
        {
            Page = users.Page,
        };
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<ExistsResponse> ExistUser(
        GuidRequired request,
        ServerCallContext context)
    {
        var doesExist = await _authorizationApp.ContainsUserAsync(request);
        var result = new ExistsResponse
        {
            DoesExist = doesExist
        };

        return result;
    }

    public override async Task<GetUserResponse> GetUser(
        GuidRequired request,
        ServerCallContext context)
    {
        var user = await _authorizationApp.GetUserAsync(request);
        var roleIds = user.RoleIds.Select(x => (GuidRequired)x);
        var result = new GetUserResponse
        {
            Id = user.Id,
            UserName = user.UserName,
            Name = user.Name,
            DisplayName = user.DisplayName,
            IsEnabled = user.IsEnabled,
        };
        result.RoleIds.AddRange(roleIds);

        return result;
    }

    public override async Task<GuidRequired> CreateUser(
        CreateUserRequest request,
        ServerCallContext context)
    {
        var command = new CreateUserCommand
        {
            UserName = request.UserName,
            Password = request.Password,
            Name = request.Name,
            DisplayName = request.DisplayName,
            IsEnabled = request.IsEnabled,
            RoleIds = request.RoleIds
                .Select(x => (Guid)x)
                .ToList(),
        };
        var result = await _authorizationApp.CreateUserAsync(command);

        return result;
    }

    public override async Task<Empty> UpdateUser(
        UpdateUserRequest request,
        ServerCallContext context)
    {
        var command = new UpdateUserCommand
        {
            Id = request.Id,
            Password = request.Password,
            Name = request.Name,
            DisplayName = request.DisplayName,
            IsEnabled = request.IsEnabled,
            RoleIds = request.RoleIds
                .Select(x => (Guid)x)
                .ToList(),
        };
        await _authorizationApp.UpdateUserAsync(command);

        return new Empty();
    }

    public override async Task<Empty> DeleteUser(
        GuidRequired request,
        ServerCallContext context)
    {
        await _authorizationApp.DeleteUserAsync(request);

        return new Empty();
    }

    public override async Task<ListNamedEntityResponse> ListRoles(
        ListRolesRequest request,
        ServerCallContext context)
    {
        var roles = await _authorizationApp.GetRolesAsync();
        var items = roles.Select(x => new ListNamedEntityResponse.Types.NamedEntity
        {
            Id = x.Id,
            Name = x.Name,
        });
        var result = new ListNamedEntityResponse();
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<PaginateRolesResponse> PaginateRoles(
        PaginateRolesRequest request,
        ServerCallContext context)
    {
        var options = new RoleOptions
        {
            Page = request.Page,
            Name = request.Name,
            IsEnabled = request.IsEnabled,
        };
        var roles = await _authorizationApp.GetRolesAsync(options);
        var items = roles.Items.Select(x => new PaginateRolesResponse.Types.Role
        {
            Id = x.Id,
            Name = x.Name,
            IsEnabled = x.IsEnabled,
        });
        var result = new PaginateRolesResponse
        {
            Page = roles.Page,
        };
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<GetRoleResponse> GetRole(
        GuidRequired request,
        ServerCallContext context)
    {
        var role = await _authorizationApp.GetRoleAsync(request);
        var permissionIds = role.PermissionIds.Select(x => (GuidRequired)x);
        var result = new GetRoleResponse
        {
            Id = role.Id,
            Name = role.Name,
            IsEnabled = role.IsEnabled,
        };
        result.PermissionIds.AddRange(permissionIds);

        return result;
    }

    public override async Task<GuidRequired> CreateRole(
        CreateRoleRequest request,
        ServerCallContext context)
    {
        var command = new CreateRoleCommand
        {
            Name = request.Name,
            IsEnabled = request.IsEnabled,
            PermissionIds = request.PermissionIds
                .Select(x => (Guid)x)
                .ToList(),
        };
        var result = await _authorizationApp.CreateRoleAsync(command);

        return result;
    }

    public override async Task<Empty> UpdateRole(
        UpdateRoleRequest request,
        ServerCallContext context)
    {
        var command = new UpdateRoleCommand
        {
            Id = request.Id,
            Name = request.Name,
            IsEnabled = request.IsEnabled,
            PermissionIds = request.PermissionIds
                .Select(x => (Guid)x)
                .ToList(),
        };
        await _authorizationApp.UpdateRoleAsync(command);

        return new Empty();
    }

    public override async Task<Empty> DeleteRole(
        GuidRequired request,
        ServerCallContext context)
    {
        await _authorizationApp.DeleteRoleAsync(request);

        return new Empty();
    }

    public override async Task<ListNamedEntityResponse> ListPermissions(
        ListPermissionsRequest request,
        ServerCallContext context)
    {
        var permissions = await _authorizationApp.GetPermissionsAsync();
        var items = permissions.Select(x => new ListNamedEntityResponse.Types.NamedEntity
        {
            Id = x.Id,
            Name = x.Name,
        });
        var result = new ListNamedEntityResponse();
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<PaginatePermissionsResponse> PaginatePermissions(
        PaginatePermissionsRequest request,
        ServerCallContext context)
    {
        var options = new PermissionOptions
        {
            Page = request.Page,
            Code = request.Code,
            Name = request.Name,
            IsEnabled = request.IsEnabled,
        };
        var permissions = await _authorizationApp.GetPermissionsAsync(options);
        var items = permissions.Items.Select(x => new PaginatePermissionsResponse.Types.Permission
        {
            Id = x.Id,
            Code = x.Code,
            Name = x.Name,
            IsEnabled = x.IsEnabled,
        });
        var result = new PaginatePermissionsResponse
        {
            Page = permissions.Page,
        };
        result.Items.AddRange(items);

        return result;
    }

    public override async Task<GetPermissionResponse> GetPermission(
        GuidRequired request,
        ServerCallContext context)
    {
        var permission = await _authorizationApp.GetPermissionAsync(request);
        var result = new GetPermissionResponse
        {
            Id = permission.Id,
            Code = permission.Code,
            Name = permission.Name,
            Description = permission.Description,
            IsEnabled = permission.IsEnabled,
        };

        return result;
    }

    public override async Task<GuidRequired> CreatePermission(
        CreatePermissionRequest request,
        ServerCallContext context)
    {
        var command = new CreatePermissionCommand
        {
            Code = request.Code,
            Name = request.Name,
            Description = request.Description,
            IsEnabled = request.IsEnabled,
        };
        var result = await _authorizationApp.CreatePermissionAsync(command);

        return result;
    }

    public override async Task<Empty> UpdatePermission(
        UpdatePermissionRequest request,
        ServerCallContext context)
    {
        var command = new UpdatePermissionCommand
        {
            Id = request.Id,
            Code = request.Code,
            Name = request.Name,
            Description = request.Description,
            IsEnabled = request.IsEnabled,
        };
        await _authorizationApp.UpdatePermissionAsync(command);

        return new Empty();
    }

    public override async Task<Empty> DeletePermission(
        GuidRequired request,
        ServerCallContext context)
    {
        await _authorizationApp.DeletePermissionAsync(request);

        return new Empty();
    }
}

File: src/Pudicitia.Enterprise.Web/src/app/management/authorization/authorization.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Guid } from 'shared/models/guid.model';
import { PaginationOutput } from 'shared/models/pagination-output.model';

import { Permission } from './permission.model';
import { RoleOutput } from './role-output.model';
import { Role } from './role.model';
import { UserOutput } from './user-output.model';
import { User } from './user.model';

@Injectable({
  providedIn: 'root',
})
export class AuthorizationService {
  private urlUsers = 'api/authorization/users';
  private urlRoles = 'api/authorization/roles';
  private urlPermissions = 'api/authorization/permissions';

  constructor(private httpClient: HttpClient) { }

  getUsers(
    page: { pageIndex: number, pageSize: number },
    userName?: string,
    name?: string,
    isEnabled?: boolean,
  ): Observable<PaginationOutput<User>> {
    let params = new HttpParams({
      fromObject: page,
    });
    if (userName !== undefined) {
      params = params.set('userName', userName!);
    }
    if (name !== undefined) {
      params = params.set('name', name!);
    }
    if (isEnabled !== undefined) {
      params = params.set('isEnabled', isEnabled!);
    }
    return this.httpClient.get<PaginationOutput<User>>(this.urlUsers, { params });
  }

  getNewUser(): Observable<UserOutput> {
    return this.httpClient.get<UserOutput>(`${this.urlUsers}/new`);
  }

  getUser(id: Guid): Observable<UserOutput> {
    return this.httpClient.get<UserOutput>(`${this.urlUsers}/${id}`);
  }

  createUser(user: User): Observable<User> {
    return this.httpClient.post<User>(this.urlUsers, user);
  }

  updateUser(user: User): Observable<User> {
    return this.httpClient.put<User>(`${this.urlUsers}/${user.id}`, user);
  }

  deleteUser(user: User): Observable<User> {
    return this.httpClient.delete<User>(`${this.urlUsers}/${user.id}`);
  }

  getRoles(
    page: { pageIndex: number, pageSize: number },
    name?: string,
    isEnabled?: boolean,
  ): Observable<PaginationOutput<Role>> {
    let params = new HttpParams({
      fromObject: page,
    });
    if (name !== undefined) {
      params = params.set('name', name!);
    }
    if (isEnabled !== undefined) {
      params = params.set('isEnabled', isEnabled!);
    }
    return this.httpClient.get<PaginationOutput<Role>>(this.urlRoles, { params });
  }

  getNewRole(): Observable<RoleOutput> {
    return this.httpClient.get<RoleOutput>(`${this.urlRoles}/new`);
  }

  getRole(id: Guid): Observable<RoleOutput> {
    return this.httpClient.get<RoleOutput>(`${this.urlRoles}/${id}`);
  }

  createRole(role: Role): Observable<Role> {
    return this.httpClient.post<Role>(this.urlRoles, role);
  }

  updateRole(role: Role): Observable<Role> {
    return this.httpClient.put<Role>(`${this.urlRoles}/${role.id}`, role);
  }

  deleteRole(role: Role): Observable<Role> {
    return this.httpClient.delete<Role>(`${this.urlRoles}/${role.id}`);
  }

  getPermissions(
    page: { pageIndex: number, pageSize: number },
    code?: string,
    name?: string,
    isEnabled?: boolean,
  ): Observable<PaginationOutput<Permission>> {
    let params = new HttpParams({
      fromObject: page,
    });
    if (code !== undefined) {
      params = params.set('code', code!);
    }
    if (name !== undefined) {
      params = params.set('name', name!);
    }
    if (isEnabled !== undefined) {
      params = params.set('isEnabled', isEnabled!);
    }
    return this.httpClient.get<PaginationOutput<Permission>>(this.urlPermissions, { params });
  }

  getNewPermission(): Observable<Permission> {
    return this.httpClient.get<Permission>(`${this.urlPermissions}/new`);
  }

  getPermission(id: Guid): Observable<Permission> {
    return this.httpClient.get<Permission>(`${this.urlPermissions}/${id}`);
  }

  createPermission(permission: Permission): Observable<Permission> {
    return this.httpClient.post<Permission>(this.urlPermissions, permission);
  }

  updatePermission(permission: Permission): Observable<Permission> {
    return this.httpClient.put<Permission>(`${this.urlPermissions}/${permission.id}`, permission);
  }

  deletePermission(permission: Permission): Observable<Permission> {
    return this.httpClient.delete<Permission>(`${this.urlPermissions}/${permission.id}`);
  }
}
